salt/utils/subprocess.py

```diff
@@ -4,12 +4,53 @@
 '''
 from __future__ import absolute_import, print_function, unicode_literals
 
+import os
+import sys
 import shlex
+import logging
 import subprocess
 import threading
 import salt.exceptions
 import salt.utils.data
 from salt.ext import six
+from salt.ext.six.moves import range
+
+
+log = logging.getLogger(__name__)
+
+
+class FdPopen(subprocess.Popen):
+    '''
+    Python2 closing file descriptors in a blind way by a system's given range.
+    Therefore close_fds can significantly slow down the system in various circumstances.
+    '''
+
+    if sys.platform.startswith('linux') and six.PY2:
+        def _close_fds(self, but):
+            '''
+            Close only file descriptors that are needed.
+            The issue is not new: https://bugs.python.org/issue1663329
+
+            :param but:
+            :return:
+            '''
+            proc_path = '/proc/{}/fd'.format(os.getpid())
+            try:
+                maxfds = len(os.listdir(proc_path))
```

Comment:
`maxfds` calculated in this way will not represent the higher opened fd number as it's desired if I'm not wrong. To do it you have to look for max integer in the list. Simple example from a live Chromium process from my PC:
```
>>> ls = os.listdir('/proc/31554/fd'); len(ls); max([int(i) for i in ls])
157
173
```

Reply:
This is a good point.

Reply:
No, this is not correct. Your Chrome is _currently_ has opened 157 files, while overall it _was opened_ 173. We are not interested how many files some session had opened in its life history, but we are only interested what we have to close after quit. So my code is correct here.

Reply:
@DmitryKuzmenko I double-checked, I think we are both wrong: you will be hammering extra-fd's that are not there anymore, and I will be leaving some (which is not entirely bad, because kernel will kill them anyways). The correct fix is to ditch the `closerange` entirely and simply iterate over, like so:

```python
for fd in [int(x) for x in os.listdir(proc_path)]:
    try:
        os.close(fd)
    ...
```

Reply:
@isbm I took your function `_close_fds` and wrote a test for it. The test results shows that I'm right here. Sorry.

The test:

```python
#!/bin/env python

import os
import subprocess


def _close_fds(but):
    '''
    Close only file descriptors that are needed.
    The issue is not new: https://bugs.python.org/issue1663329

    :param but:
    :return:
    '''
    proc_path = '/proc/{}/fd'.format(os.getpid())
    try:
        maxfds = len(os.listdir(proc_path))
    except OSError:
        maxfds = subprocess.MAXFD

    print('Closing file descriptors from 3 to {}'.format(maxfds))
    if hasattr(os, 'closerange'):
        if 3 <= but <= maxfds:
            os.closerange(3, but)
            os.closerange(but + 1, maxfds)
        else:
            os.closerange(3, maxfds)
    else:
        for i in range(3, maxfds):
            if i != but:
                try:
                    os.close(i)
                except Exception:
                    pass


proc_path = '/proc/{}/fd'.format(os.getpid())
print('On start:           {}'.format(os.listdir(proc_path)))
files = []
for i in range(10):
    files.append(open('/tmp/f{}'.format(i), 'w'))
print('Opened 10 files:    {}'.format(os.listdir(proc_path)))
for i in range(5):
    print('Closing fd#{}'.format(files[i].fileno()))
    files[i].close()
print('After closing 3-7:  {}'.format(os.listdir(proc_path)))
_close_fds(0)
print('After _close_fds(): {}'.format(os.listdir(proc_path)))
```

Test output:

```
$ python test.py 
On start:           ['0', '1', '2', '3']
Opened 10 files:    ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13']
Closing fd#3
Closing fd#4
Closing fd#5
Closing fd#6
Closing fd#7
After closing:      ['0', '1', '2', '3', '8', '9', '10', '11', '12']
Closing file descriptors from 3 to 9
After _close_fds(): ['0', '1', '2', '3', '9', '10', '11', '12']
```

As you can see after opening 10 files we have 13 files opened (14th is opened by `os.listdir` just while `os.listdir` works, I don't calc it here). After I close 5 files in the middle of this range we have 8. `_close_fds` tries to close range 3-9 (9 is excluded by `closerange` by design). So we have 9 to 12 opened as a result.

Also please take note that I've fixed how `_close_fds` handles `but` in case it use `closerange`. It's needed if `0` is passed meaning "no but, close all". In your implementation it closes 0-2 descriptors too if `0` passed as `but`.

Reply:
@isbm didn't see your comment. Yes, that is the best and simplest way as I see.

Reply:
@DmitryKuzmenko I found exactly the same thing. :wink: So thanks for double-checking this as well. I forgot that `os.close()` is by descriptor num, and hence we need exact ones.

Reply:
Why not reduce the file limit.    ulimit setting being high and MAX_OPEN_FILES being high is only required for network connections.   Why not have the rest of salt stay at 256 fds and only the process which accepts connections have higher connection count. Might need to use dup2.  Have not looked at how salt handles in coming connections.